### PR TITLE
Add singly-linked implementation of immutable stacks.

### PR DESCRIPTION
**Pull request by motlin**
Signed-off-by: Craig P. Motlin <cmotlin@gmail.com>

### CODE REVIEW

**Comment by motlin** [comment]
While this change works, I think it may require some discussion. The way I wrote it, `org.eclipse.collections.impl.stack.immutable.ImmutableArrayStack` is basically dead code.

**Comment by nikhilnanivadekar** [comment]
I missed doing a review of this. Sorry about that. I will check this later today.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/stack/immutable/ImmutableNotEmptyStack.java [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+        return this.element;
+    }
+
+    @Override
+    public ListIterable<T> peek(int count)
+    {
+        this.checkNegativeCount(count);
+        if (this.checkZeroCount(count))
+        {
+            return FastList.newList();
```
Comment (donraab) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/stack/immutable/ImmutableNotEmptyStack.java:138]: use `Lists.mutable.empty()`

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/stack/immutable/ImmutableNotEmptyStack.java [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+        }
+        this.checkSizeLessThanCount(count);
+        return FastList.newList(this.asLazy().take(count));
```
Comment (donraab) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/stack/immutable/ImmutableNotEmptyStack.java:141]: Can this be `this.asLazy().take(count).toList()`?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/stack/immutable/ImmutableNotEmptyStack.java (hunk 1 above)
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/stack/immutable/ImmutableNotEmptyStack.java]: Done

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/stack/immutable/ImmutableNotEmptyStack.java (hunk 2 above)
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/stack/immutable/ImmutableNotEmptyStack.java]: Yes, done

**Pull request merged**